src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,87 @@
+import java.util.HashMap;
+import java.util.Map;
+import mate.academy.service.FileReaderService;
+import mate.academy.service.ProductParser;
+import mate.academy.service.ProductService;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private static final Map<Class<?>, Object> instances = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        if (interfaceClazz == null) {
+            throw new RuntimeException("Class or interface is null");
+        }
+        Object clazzImplementationInstance = null;
+        Class<?> clazz = findImplementation(interfaceClazz);
+        if (!clazz.isAnnotationPresent(Component.class)) {
+            throw new RuntimeException("This class does not have component annotation");
+        }
+        Field[] fields = interfaceClazz.getFields();
+        for (Field field : fields) {
+            if (field.isAnnotationPresent(Inject.class)) {
+                Object fieldInstance = field.getType();
+                clazzImplementationInstance = createNewInstance(clazz);
+                field.setAccessible(true);
+                try {
+                    field.set(clazzImplementationInstance, fieldInstance);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Can't initialize fields value. Class "
+                            + clazz.getName() + ". field " + field.getName(), e);
+                }
+            }
+        }
+        if (clazzImplementationInstance == null) {
+            clazzImplementationInstance = createNewInstance(clazz);
+        }
+        return clazzImplementationInstance;
+    }
+
+    private Object createNewInstance(Class<?> clazz) {
+        if (clazz == null) {
+            throw new RuntimeException("Class or interface is null");
+        }
+        if (instances.containsKey(clazz)) {
+            return instances.get(clazz);
+        }
+        Constructor<?> constructor;
+        try {
+            constructor = clazz.getConstructor();
+            Object object = constructor.newInstance();
+            instances.put(clazz, object);
+            return object;
+
+        } catch (NoSuchMethodException | InvocationTargetException
+                 | IllegalAccessException | InstantiationException e) {
+            throw new RuntimeException("Can't create a new instance of "
+                    + clazz.getName(), e);
+        }
+    }
+
+    private Class<?> findImplementation(Class<?> interfaceClazz) {
+        if (interfaceClazz == null) {
+            throw new RuntimeException("Class or interface is null");
+        }
+        Map<Class<?>, Class<?>> implementationsMap = new HashMap<>();
```

Comment:
implementationsMap could be an instance variable and filled using method Map.of() for better readability.